Update plugin path shortening logic

rdswift opened the pull request:
# Summary

* This is a…
  * [ ] Bug fix
  * [ ] Feature addition
  * [ ] Refactoring
  * [ ] Minor / simple change (like a typo)
  * [x] Other
* **Describe this change in 1-2 sentences**:

# Problem

Further to the discussion in https://github.com/metabrainz/picard/pull/2397 this PR modifies the way that plugin paths are displayed in the logs.

* JIRA ticket (_optional_): PICARD-XXX

# Solution

Shorten the path by making it relative to the path stored in the `picard.const.USER_PLUGIN_DIR` variable if possible.

# Action

None.

zas commented on picard/log.py:
```diff
@@ -177,11 +179,15 @@ def name_filter(record):
+        try:
+            path = path.resolve().relative_to(USER_PLUGIN_DIR)
+            parts = list(p for p in path.parts if not p.endswith('.zip'))
+            parts.insert(0, 'plugins')
```
I don't understand this insert.


```python
>>> path = Path('/a/b/c/subdir/e/f.zip/g')
>>> USER_PLUGIN_DIR = Path('/a/b/c/subdir')
>>> path = path.resolve().relative_to(USER_PLUGIN_DIR)
>>> str(path)
'e/f.zip/g'
>>> parts = list(p for p in path.parts if not p.endswith('.zip'))
>>> parts
['e', 'g'] 
>>> parts.insert(0, 'plugins')
>>> parts
['plugins', 'e', 'g']
>>> path = Path(*parts)
>>> str(path)
'plugins/e/g'
```

Why would we introduce an element which isn't in any of existing paths?

rdswift replied:
> Why would we introduce an element which isn't in any of existing paths?

It sort of is in the original `USER_PLUGIN_DIR` in Picard, but gets stripped out when the path is resolved relative to `USER_PLUGIN_DIR`.  My thinking was to add it back in to clearly indicate in the log that this was in a plugin.

I'll remove it.

zas commented on test/test_log.py:
```diff
@@ -162,21 +164,34 @@ def test_5(self):
+        DebugOpt.PLUGIN_FULLPATH.enabled = True
+        record = FakeRecord(name=None, pathname='/user/picard/plugins/plugin.zip/plugin')
+        self.assertTrue(name_filter(record))
+        self.assertEqual(record.name, 'user/picard/plugins/plugin.zip/plugin')
```
Why are we ending with a path not starting with `/` here?

rdswift replied:
I believe that's because of stripping out the anchor in the lines in `log.py`:

```python
    if parts[0] == path.anchor:
        del parts[0]
```

That's the same behavior we're seeing when the path isn't relative to the `picard` module.  For example, in `test_log.py`:

```python
@unittest.skipIf(IS_WIN, "Posix test")
@patch('picard.log.picard_module_path', PurePosixPath('/picard'))
class NameFilterTestAbs(PicardTestCase):


    def test_1(self):
        record = FakeRecord(name=None, pathname='/path/module/file.py')
        self.assertTrue(name_filter(record))
        self.assertEqual(record.name, 'path/module/file')
```

If you want the absolute (non-shortened) paths to start with a "/" (for both Windows and POSIX) the code could be changed to:

```python
    if parts[0] == path.anchor:
        parts[0] = '/'
```

I'll make the change, because it will make it clearer that the path is absolute.

zas commented on test/test_log.py:
```diff
@@ -162,21 +164,34 @@ def test_5(self):
+    def test_plugin_path_short_2(self):
         DebugOpt.PLUGIN_FULLPATH.enabled = False
-        record = FakeRecord(name=None, pathname='/path1/path2/plugins/path3/plugins/plugin.zip')
+        record = FakeRecord(name=None, pathname='/user/picard/plugins/plugin.zip/plugin')
```
It would be better to use `'/user/picard/plugins/plugin.zip/xxx' here because we don't know which `plugin` was kept in the result.

zas commented on test/test_log.py:
```diff
@@ -168,25 +168,25 @@ def test_plugin_path_long_1(self):
+        record = FakeRecord(name=None, pathname='/user/picard/plugins/plugin.zip/xxx')
         self.assertTrue(name_filter(record))
-        self.assertEqual(record.name, 'plugins/plugin')
+        self.assertEqual(record.name, 'xxx')
```
In this case it looks weird to lose the plugin name (from `plugin.zip`), I guess having `plugin/xxx` as result would make much more sense.

rdswift replied:
What do you think of prefacing all shortened plugin paths with "Plugin:" to clearly indicate that's what they are?  For example, "/home/user/picard/plugins/xxx.zip/yyy" would be logged as "Plugin: yyy" and "/home/user/picard/plugins/xxx.zip" would be logged as "Plugin: xxx" (assuming that `USER_PLUGIN_DIR` was "/home/user/picard/plugins").

Otherwise, I can just put it back to starting with "plugins/".  I'm okay either way.  Just let me know what you prefer.

zas replied:
Hmmm, I now get why you added `plugins/` at first. And I guess that's the best option.

rdswift replied:
Actually, in looking at the plugins, stripping out the zip file from the path might not be such a good idea because in some cases the file inside may be `__init__.py`.  Perhaps we should leave the zip file name in the path to help ensure that the plugin is properly identified.  That way, some examples might be:

* `Plugin: abbreviate_artistsort.zip/abbreviate_artistsort`
* `Plugin: add_to_collection.zip/__init__`
* `Plugin: extra/path/plugin.zip/__init__`

Or we could stop at the zip file and strip the ".zip", but that wouldn't show which file it was within the plugin (if there were multiple modules).  For example, the three above would be:

* `Plugin: abbreviate_artistsort`
* `Plugin: add_to_collection`
* `Plugin: extra/path/plugin`

Considering that we have the option to show the full path if needed, I'm actually tempted to do it this way.  What do you think?

rdswift replied:
I'll put it back to just including the "plugin/" for now and we can change later if we decide we want to go a different route.

rdswift replied:
Another option, and I think this might actually make the most sense, is to not add the "plugins" back in, but also not remove the zip file from the path.  This keeps it as a shortened (relative) path, and the presense of a zip file in the path clearly indicates that this is a plugin.  Some examples might be:

* `abbreviate_artistsort.zip/abbreviate_artistsort`
* `add_to_collection.zip/__init__`
* `some_plugin.zip/module_path/some_module`
* `extra/path/plugin.zip/__init__`

zas reviewed (approved):
LGTM

rdswift commented:
Revised the logic slightly to include the zip file in the shortened plugin path for clarity.  Also fixed some tests that I had screwed up regarding absolute path testing, and added additional tests.

zas reviewed (approved):
LGTM
